executor/executor.go

```diff
@@ -765,10 +775,26 @@ func (e *ShowDDLJobQueriesWithRangeExec) Next(ctx context.Context, req *chunk.Ch
 		return nil
 	}
 	numCurBatch := mathutil.Min(req.Capacity(), len(e.jobs)-e.cursor)
-	for i := e.cursor; i < e.cursor+numCurBatch; i++ {
-		if i >= int(e.offset) && i < int(e.offset+e.limit) {
-			req.AppendString(0, strconv.FormatInt(e.jobs[i].ID, 10))
-			req.AppendString(1, e.jobs[i].Query)
+	numOfRowsToWrite := numCurBatch
```

Comment:
now this logic is too long and readability is also not good. 

`ShowDDLJobQueriesExec` is different with `ShowDDLJobQueriesWithRangeExec`, `ShowDDLJobQueriesExec` chooses to loop a batch of rows multiple times to match `jobIDs`, so the batch size is needed to control loop granularity. `ShowDDLJobQueriesWithRangeExec` will only access once for each row, seems we can directly let `e.cursor++` and `numCurBatch--`, no need to use `i` and `numOfRowsToWrite` to keep original values

Reply:
If we directly let `e.cursor++` and `numCurBatch--`, `e.cursor` will not end up the same as when we use `numOfRowsToWrite` and `e.cursor += numCurBatch`. When we detect duplicate jobs, we need to `i++` to set the local cursor forward to check the next job, but in this case `numCurBatch` should not decrease 1 since we did not write this line. So if we directly use `e.cursor` to be `i`, when `numCurBatch` reaches 0, `e.cursor` will end up larger than it should be, which should be the original `e.cursor` plus original `numCurBatch`. Then the next query will get the wrong `e.cursor`.

Reply:
This is the code of `Next` I modified according to this suggestion, I would like to make sure that I understand you right

```
// Next implements the Executor Next interface.
func (e *ShowDDLJobQueriesWithRangeExec) Next(ctx context.Context, req *chunk.Chunk) error {
	req.GrowAndReset(e.maxChunkSize)
	if e.cursor >= len(e.jobs) {
		return nil
	}
	if int(e.limit) > len(e.jobs) {
		return nil
	}
	numCurBatch := mathutil.Min(req.Capacity(), len(e.jobs)-e.cursor)
	appendedJobID := make(map[int64]struct{})
	if e.cursor < int(e.offset) {
		e.cursor = int(e.offset)
	}
	for {
		// check whether the job has been read
		// for situations when this operation happens at the same time with new DDLs being executed
		if _, ok := appendedJobID[e.jobs[e.cursor].ID]; !ok {
			appendedJobID[e.jobs[e.cursor].ID] = struct{}{}
			req.AppendString(0, strconv.FormatInt(e.jobs[e.cursor].ID, 10))
			req.AppendString(1, e.jobs[e.cursor].Query)
			numCurBatch--
		}
		e.cursor++
		if e.cursor >= int(e.offset+e.limit) || e.cursor >= len(e.jobs) || numCurBatch == 0 {
			break
		}
	}
	return nil
}
```

Reply:
The new code is easier to understand, and sorry I just found other problems that I didn't realise in previous review 😂 :

`appendedJobID` is reset to a new variable after `Next` invocation is finished, so it can't deduplicate result in different `Next`, considering the caller will use Open-Next-Next(multiple times)-Close to get full result.

And as you says the skipped rows will cause the relation of `e.cursor`, `numCurBatch` and other variables a bit messy. Now I prefer we do the deduplication in `Open`

```
	e.jobs = append(e.jobs, jobs...)
	e.jobs = append(e.jobs, historyJobs...) <- change it to skip duplicate jobs
```

So when it goes to `Next`, we can directly start to write e.jobs[e.offset:e.offset+e.limit] in `Next` until the input chunk is full.